src/style-spec/function/evaluation_context.js

```diff
@@ -99,72 +95,107 @@ module.exports = () => ({
             const c = parseColor(input);
-            if (!c)
-                throw new RuntimeError(`Could not parse color from value '${input}'`);
-            cached = this._parseColorCache[input] = new Color(...c);
+            cached = this._parseColorCache[input] = c ? new Color(...c) : null;
         }
         return cached;
     },
 
     rgba: function (r: number, g: number, b: number, a?: number) {
-        ensure(r >= 0 && r <= 255 &&
-            g >= 0 && g <= 255 &&
-            b >= 0 && b <= 255, `Invalid rgba value [${[r, g, b, a || 1].join(', ')}]: 'r', 'g', and 'b' must be between 0 and 255.`);
-        ensure(typeof a === 'undefined' ||
-            (a >= 0 && a <= 1), `Invalid rgba value [${[r, g, b, a || 1].join(', ')}]: 'a' must be between 0 and 1.`);
+        const error = this._validateRGBA(r, g, b, a);
+        if (error) throw new RuntimeError(error);
         return new Color(r / 255, g / 255, b / 255, a);
     },
 
+    _validateRGBA(r: number, g: number, b: number, a?: number): ?string {
+        if (!(
+            typeof r === 'number' && r >= 0 && r <= 255 &&
+            typeof g === 'number' && g >= 0 && g <= 255 &&
+            typeof b === 'number' && b >= 0 && b <= 255
+        )) {
+            const value = typeof a === 'number' ? [r, g, b, a] : [r, g, b];
+            return `Invalid rgba value [${value.join(', ')}]: 'r', 'g', and 'b' must be between 0 and 255.`;
+        }
+
+        if (!(
+            typeof a === 'undefined' || (a >= 0 && a <= 1)
+        )) {
+            return `Invalid rgba value [${[r, g, b, a].join(', ')}]: 'a' must be between 0 and 1.`;
+        }
+
+        return null;
+    },
+
     toString: function(value: Value) {
-        const type = this.typeOf(value);
-        ensure(value === null || /^(String|Number|Boolean)$/.test(type), `Expected a primitive value in ["string", ...], but found ${type} instead.`);
-        return String(value);
+        const type = typeof value;
+        if (value === null || type === 'string' || type === 'number' || type === 'boolean') {
+            return String(value);
+        } else if (value instanceof Color) {
+            const [r, g, b, a] = value.value;
+            return `rgba(${r * 255}, ${g * 255}, ${b * 255}, ${a})`;
+        } else {
+            return JSON.stringify(value);
+        }
     },
 
-    toNumber: function(value: Value) {
-        const num = Number(value);
-        ensure(value !== null && !isNaN(num), `Could not convert ${JSON.stringify(this.unwrap(value))} to number.`);
-        return num;
+    toNumber: function() {
+        // use arguments object instead of rest args because buble translation
+        // of rest args allocates an array
+        let value;
+        for (let i = 0; i < arguments.length; i++) {
+            value = arguments[i];
+            if (value === null) continue;
+            const num = Number(value);
+            if (isNaN(num)) continue;
+            return num;
+        }
```

Comment:
Is there any way to avoid relying on `arguments` in a runtime-called method? Perhaps we could cache the possible values in an array like we do for stop outputs (and get rid of this method in the eval context)? Let's check and maybe benchmark to be sure it's faster.

Reply:
👍, good call, will try that out.

@mourner btw, I was going off of the rule of thumb that using `arguments` is fine if you never use the object itself, only access `arguments.length` and `arguments[i]` -- what's your take on that?

Reply:
@mourner fixed -- perf was equivalent, this way seems cleaner/better anyway